### PR TITLE
Suggested improvements to testing on a k8s cloud using fixtures

### PR DESCRIPTION
**Pull request by addyess**
@sachinkumarsingh092 Here are some suggestions to testing coredns in microk8s

Please see this PR raises into your `dev/kube-dns` branch and not `master`

### CODE REVIEW

File: tests/integration/conftest.py [hunk 1]
```diff
@@ -107,12 +107,8 @@ async def microk8s_kubernetes_cloud_and_model(ops_test: OpsTest, microk8s_applic
         yield ("k8s-model", model_name)
 
     finally:
-        await ops_test.forget_model("k8s-model")
         LOG.info("Destroy model %s", model_name)
-        res = await ops_test.juju(
-            "destroy-model", model_name, "--force", "--destroy-storage", "--yes", "--no-prompt"
-        )
-        LOG.info("%s", res)
+        await ops_test.forget_model("k8s-model", destroy_storage=True)
```
Comment (addyess) [inline tests/integration/conftest.py:111]: this is basically doing the same thing twice.

File: tests/integration/test_coredns.py [hunk 2]
```diff
@@ -1,23 +1,21 @@
...
-from juju.application import Application
 from pytest_operator.plugin import OpsTest
 
 LOG = logging.getLogger(__name__)
 
 
-@pytest.mark.abort_on_fail
-async def test_core_dns(e: OpsTest):
+@pytest.fixture()
+async def microk8s(e: OpsTest):
```
Comment (addyess) [inline tests/integration/test_coredns.py:18]: deploy the microk8s charm as a fixture returning the active/idle `Application`

File: tests/integration/test_coredns.py [hunk 3]
```diff
@@ -29,41 +27,55 @@ async def test_core_dns(e: OpsTest):
         )
         await e.model.wait_for_idle(["microk8s"])
 
-    u: Unit = e.model.applications["microk8s"].units[0]
+    yield e.model.applications["microk8s"]
+
 
+@pytest.fixture()
+async def coredns(e: OpsTest, microk8s):
```
Comment (addyess) [inline tests/integration/test_coredns.py:34]: deploy the `coredns` charm as an `active/idle` fixture and the model with the offers available to link the machine model. 

on teardown -- remove the offer and the application

File: tests/integration/test_coredns.py [hunk 4]
```diff
@@ -29,41 +27,55 @@ async def test_core_dns(e: OpsTest):
...
 
-            with e.model_context(k8s_model):
-                await e.model.wait_for_idle(["coredns"])
 
-            rc = None
-            while rc != 0:
-                rc, stdout, stderr = await run_unit(u, f"microk8s kubectl run -i --tty --rm debug --image=busybox --restart=Never -- nslookup google.com")
-                LOG.info("Verify the pod dns resolution %s", (rc, stdout, stderr))
+@pytest.mark.abort_on_fail
+async def test_core_dns(e: OpsTest, microk8s, coredns):
```
Comment (addyess) [inline tests/integration/test_coredns.py:56]: test by consuming the offer in the machine_model, waiting for both models to be idle, and running the dns query

then tearing out the relation and saas

File: tests/integration/test_coredns.py [hunk 5]
```diff
@@ -29,41 +28,62 @@ async def test_core_dns(e: OpsTest):
...
+            await k8s_model.wait_for_idle(["coredns"])
+
+        yield k8s_model.applications["coredns"]
+
+        k8s_model.remove_offer("coredns")
+        k8s_model.remove_application("coredns", block_until_done=True)
+
+
+@pytest_asyncio.fixture()
+async def related_microk8s_coredns(microk8s, coredns):
```
Comment (addyess) [inline tests/integration/test_coredns.py:57]: Relate microk8s and coredns using a fixture

File: tests/integration/conftest.py [hunk 6]
```diff
@@ -107,12 +109,18 @@ async def microk8s_kubernetes_cloud_and_model(ops_test: OpsTest, microk8s_applic
...
+        await ops_test.forget_model("k8s-model", destroy_storage=True, timeout=timeout)
+
+        async def model_removed():
+            rc, stdout, _ = await ops_test.juju("models", "--format", "yaml")
+            if rc != 0:
+                return False
+            model_list = yaml.safe_load(stdout)["models"]
+            return len([m for m in model_list if m["name"] == f"admin/{model_name}"]) == 0
+
+        await block_until_with_coroutine(model_removed, timeout=timeout)
```
Comment (addyess) [inline tests/integration/conftest.py:123]: don't delete the connected cloud until the its models are removed

**Comment by sachinkumarsingh092** [comment]
Thanks for all the great suggestions @addyess.

**Pull request merged**